Add RandomGrid image perturber

schencej opened the pull request.

Purg commented:
Rebased on updated master

Purg commented on xaitk_saliency/impls/perturb_image/square_grid.py:
```diff
@@ -0,0 +1,82 @@
+
+        img_shape = np.array(ref_img.shape[:2])
+
+        grid_size = (img_shape // s) + 1
+        grid_size += img_shape % s != 0
```
Things seem to work, but this math threw me for a loop as to why the `+1` is here, etc. A comment or two might address why this is being done.

schencej commented:
- Added docstring and comments
- Changed the `s` parameter to be a tuple, requiring the user to specify the cell size along both dimensions. Also changed the class name to a more generic `RandomGrid` since it is not explicitly square now
 - If you used a cell size of 1, the max shift was 0 which caused an error to be thrown by the random number generator. I added logic for this

schencej commented:
Is mentioning RISE and/or including its license in the docstring something we should do here? The logic is similar.

Purg commented on xaitk_saliency/impls/perturb_image/random_grid.py:
```diff
@@ -0,0 +1,118 @@
+        seed: Optional[int] = None,
+        threads: Optional[int] = None,
+    ):
+        """
+        Generate masks using a random grid of set cell size. If the chosen cell
+        size does not divide an image evenly, then the grid is over-sized and
+        the resulting mask is centered and cropped. Each mask is also shifted
+        randomly by a maximum of half the cell size in both x and y.
+
+        :param n: Number of masks to generate.
+        :param s: Dimensions of the grid cells in pixels. E.g. (3, 4) would use
+            a grid of 3x4 pixel cells.
+        :param p: Probability of a grid cell being set to 1 (not occluded).
+            This should be a float value in the [0, 1] range.
+        :param seed: A seed to use for the random number generator, allowing
+            for masks to be reproduced.
+        :param threads: Number of threads to use when generating masks. If this
+            is <=0 or None, no threading is used and processing is performed
+            in-line serially.
+        """
```
Somewhat minor, but please move this doc-string to the class level.

Purg commented:
> Is mentioning RISE and/or including its license in the docstring something we should do here? The logic is similar.

We could mention that this impl is based on RISE and exists to address the changing-"square"-size aspect of the original RISE implementation. I think it is important to capture where we came from (RISE) and why we are here (fixed pixel rectangles).

Purg reviewed (approved):
LGTM.

brianhhu commented on xaitk_saliency/impls/perturb_image/random_grid.py:
```diff
@@ -0,0 +1,124 @@
+    :param n: Number of masks to generate.
+    :param s: Dimensions of the grid cells in pixels. E.g. (3, 4) would use
+        a grid of 3x4 pixel cells.
+    :param p: Probability of a grid cell being set to 1 (not occluded).
```
Quick question- should we rename this parameter `p1` (here and below) to be consistent with RISE? I had some confusion whether the probability `p` corresponded to setting a grid to 0 or 1 (in some earlier discussions with Elim), and adding the 1 may help resolve some of this ambiguity.

schencej replied:
I think that makes sense. I'll update and autosquash.

brianhhu reviewed (commented):
Can we also add this new implementation to our docs [here]?

brianhhu commented:
LGTM, thanks!
